### ISSUE TITLE
Automatic resuming of flows after azkaban-executor restart

### ISSUE THREAD

**Issue by juhoautio** [azkaban/azkaban#1127]
This is a new feature suggestion. Feedback much appreciated!

## Rationale

We need to support resuming flows and jobs after azkaban-executor is restarted. Currently Azkaban just marks interrupted flows and jobs as FAILED on startup. This is actually done silently without sending flow failure notifications or anything like that.

Reasons for "restarting" azkaban-executor include at least:
- Azkaban-executor java process crashed (OOM, for example)
- Azkaban version update requires a restart to apply changes (especially in a single executor or solo-azkaban setup)
- Hardware failure or some other problem requires replacing the azkaban-executor with an entirely new instance

## Design

Actions on startup for interrupted flows:
- Start execution for the flow again with the _same execution id and project version_
  - For example flow log should have all the history before the flow was interrupted + note that execution was resumed
- Jobs that support resume are started again with `JobState.isResume=true`
- Jobs that don't support resume are started over
  - So jobs that don't support resume get "1 extra retry" by default
  - This can be disabled by setting `resume.retryJobIfResumeNotImplemented=false`
- Jobs are resumed without incrementing the retry attempt count – they should have the attempt number that was running when flow was interrupted
- Add new flow level email notification option `resume.emails` (similar to the existing `failure.emails`)

`Job` API change:
- New optional method: `boolean Job#supportsResume`, default: `false`
- New optional method: `void Job#setState(JobState)` (this is not called unless `supportsResume` returns `true`)

New interface `JobState` with methods:
- `void JobState#setState(Props)`
  - This persists the given state `Props` into DB (key: flow/execution/job/attempt)
- `Props JobState#getState`
- `boolean JobState#isResume`
  - Jobs could basically get the same result by checking `JobState.getState() != null` in the beginning of the run, but I think that having a specific method is useful.
  - Naturally `isResume` will keep returning `false` for a regular (not resumed) execution even if the job saves something to the `JobState`.

This way the job can store some execution state that is fetched back for it before resuming.

## Notes about our use cases

Usually we just need to store an external job id after it has been launched, and when resuming we continue polling for the external status by id.

Resuming can also be useful for "locally running jobs" if they know how to continue from whatever state they have saved so far.

In practice we for example launch AWS EMR clusters/jobs externally with a custom job type for that purpose. We have some other custom job types as well that launch a job externally and then poll for that to complete.

In most of the cases we would need to store an external job id after it has been launched. When resuming we either continue polling for the external status by id, or launch a new external job if there's no existing job id to track.

**Comment by goelrajat** [comment]
Any updates on when this feature will be available?

**Comment by juhoautio** [comment]
Hopefully soon, but as you can see this post is only 4 days old. First we would need review of this plan by repo owners and an "ok" before we can start implementing it in practice. I would say it will still take weeks before this could be seen as a working feature in the master branch.

**Comment by HappyRay** [comment]
Thanks for the detailed write-up.
I will have to think more about it.

At Linkedin, we currently only start a new executor process and let the existing flows finish before shutting down the old executor processes. We don't have any resumable job type currently. 
Yes, it wold be nice to support resuming in a case of an H/W failure. However, h/w failure is rare so far. OOM in the executor process is also rare. I am interested in supporting a way to restart the executor process without affecting the running job processes. This will simplify our deployment process and gives us the flexibility to recover from a software bug in the executor more quickly. 

Do h/w failure and OOM happen often in your environment? Or is the ease of software update the main motivation for you?

**Comment by juhoautio** [comment]
The main motivation is being able to ensure that important scheduled flows are completed even if the executor crashes in the middle of an execution. A part of ensuring that is the ability to continue externally running jobs in a clean way after any kind of restart.
- It would be a real problem if we couldn't continue polling externally launched jobs after a restart and launch a new overlapping job instead. Currently we have a custom way to synchronize the external jobs via a DB, but this still means that when we resume polling the job gets a new execution id and the previous execution has been marked as FAILED.

Main OOM scenarios are:

1. Metaspace (Java 8) runs out after reloading custom job types multiple times
    - Because each reload creates new classloaders that aren't always entirely cleaned up, most often because of some background threads left running and preventing classes to be garbage collected from the old classloaders
1. OS level "OOM" that makes OS sacrifice java process and kills it
    - This can happen if additional processes created by command jobs consume too much memory (as a special case when a process creates sub-processes). That's partially caused because we have disabled available memory check before running command jobs, because of _reasons_.
1. Too many / heavy flows running at the same time, basic OOM
    - I can't be sure, but I think this is because azkaban-executor creates instances of ExcutableNode/ExecutableFlow for each node in the graph.
    - Example of heavy graph: you have chained, say, some big flow, into a bigger flow that runs the same big sub-flow in a of sequence 30 times. 
1. Running out of heap size because of running enough ~memory-hungry jobs that run within the executor JVM
    - Hasn't really happened to us, but I can imagine it rather easily

Of course if we hit any of these issues we try to fix the cause of the problem. But you can't always predict problems so Azkaban should be able to recover from crashes automatically.

**Comment by chengren311** [comment]
@juhoautio i think it's a valid feature request. Thanks! I'm curious about how you would plan to implement setState and getState, and how do we define a state exactly?

**Comment by juhoautio** [comment]
> I'm curious about how you would plan to implement setState and getState

First of all note that `Job#setState(JobState)` and `JobState#setState(Props)` are entirely different methods. Actually should rename the latter to `JobState#setProps(Props)`. But I'll stick with the original method name for now..

Then to some kind of impl. plan.

Look at [JobTypeManager#buildJobExecutor](https://github.com/azkaban/azkaban/blob/fc5d415635e2d10f972538c4c7d85b50db316cb0/azkaban-common/src/main/java/azkaban/jobtype/JobTypeManager.java#L311). When building the executor:
1. Check if the job is resumable (how exactly, depends on what kind of Job API is chosen for resumable job types)
1. Instantiate the job type as usual (call constructor with reflection)
1. Call `Job#setState` on the created instance
1. Continue as normal, eventually `Job#run` is called

Creating the `JobState`:
1. This would probably happen elsewhere and get passed as an argument to `JobTypeManager#buildJobExecutor`
1. Load the state `Props` from DB by `<execution id + attempt number>` (if exists).
1. If state is found for this execution/attempt, then it's a resume case and `JobState` is created with those `Props` and `isResume=true`. Otherwise it's created with empty `Props` and `isResume=false`.
1. Set the state on the job by calling `Job#setState`

Also `JobState` needs to have a way to persist the `Props` into DB when the Job calls `JobState #setState`. So some kind of DB gateway instance must be passed to `JobState` constructor.

> and how do we define a state exactly?

I'm not sure what this question means?

**Comment by priyanshjain28** [comment]
@juhoautio Hi, this is exactly what we are looking forward to do. We process data in our jobs scheduled in azkaban. If azkaban goes down and comes up after a certain period of time, there is a data hole created for the jobs which did not run. Below example might clear this up:

Case Scenario: A job is scheduled hourly in azkaban. Let's say the job takes 45 minutes for successful completion. Let the scheduled time for the job be:
12pm -> 1pm -> 2pm -> 3pm -> 4pm and so on..
Let's say azkaban went down at 12:30pm and came up at 2:30pm.

Current Scenario: Azkaban will not restart the failed job (12pm job) or any missing instances in between. It will continue with its execution by running the next scheduled instance i.e. 3pm.

Expected Scenario: Azkaban should run in replay mode and re-run the 12pm job and other missing jobs. Here we have 2 cases:
1. Job was scheduled keeping concurrency on: Azkaban will launch all the missing instances together. Thus, at 2:30pm, three instances correponding to 12pm, 1pm and 2pm will be launched concurrently. Azkaban will continue with normal execution post this.
2. Job was scheduled keeping pipeline feature on: Azkaban will launch the 12pm instance at 2:30pm. It will put other instances in a pipeline. Once the execution of 12pm instance(which was re-run at 2:30pm) is over at 3:15(45min is the processing time for one job), azkaban will launch the 1pm instance at 3:15pm. This way, azkaban will try to catch up the lag period. Once azkaban catches up, normal hourly execution would continue.


Also, if possible, could you let us know a time-period by which this feature might be available? Or if there has been any progress at your end? We can probably discuss the design, if you have one ready.
Because we are also planning on making the changes in source code for the same. Just want to avoid duplicate effort. 
Any help would be appreciated. :)

**Comment by juhoautio** [comment]
Thanks @priyanshjain28. We didn't consider "replay" mode for schedules, but that's a good addition, definitely. I think it could be handled separately, though. We haven't written any code for this yet. I was aiming at getting a common agreement on the specification with the repo owners before starting to implement anything. The goal being, getting code review to catch problems, and having this feature merged into the public repo to avoid diff with our own fork. We could have time to implement this, but I'm not sure how quickly we will be able to have a spec greenlighted by @HappyRay and @chengren311.

**Comment by juhoautio** [comment]
I'm going to start working on this. As there hasn't been a chance to prepare & agree on the spec with repo owners, I believe it won't be easy to get it merged in.. But even if so, please give any last feedback you might have. I think I'm going to drop a PR just for reference once this feature is ready.

**Comment by HappyRay** [comment]
@juhoautio sorry about the delay. 
I plan to refactor the core DAG processing code to make it easier to maintain e.g. easier to unit test and understand. I hope to find time to do so in the next month or so. 
This will likely touch many classes. I hope the end result will be better code that is easier to add new features.
If you can wait until the big refactoring is done, it may make your job and merge easier. 

For your use case:
> Azkaban version update requires a restart to apply changes (especially in a single executor or solo-azkaban setup)
Do you run in a multiple executor mode? If so, you should be able to start a new executor process to replace the existing instance as an active executor. Let the existing one finish processing existing flows. It has worked for us. Would it work for you?

**Comment by juhoautio** [comment]
I expect resuming functionality to have no impact on the DAG execution part of the application. Or it should be small at least.

We don't run multi-executor yet, but hoping to switch to that at some point. Making that switch requires some additional tooling and conf changes.

Yes, using multi-executor mode would help us with a deployment a bit and that alone is a reason enough to take it into use. But it won't help us with:
- executor crash scenarios (because azkaban doesn't start the interrupted flows again)
- long-running "streaming jobs" (can be running for days) that are running on an external cluster, that we'd rather not interrupt, but just continue tracking them

**Comment by swamisun** [comment]
@juhoautio thanks for initiating this. Ability for flows to withstand (rolling) restarts would be great to have, and we'd love to help realize this. Any progress more progress on this besides https://github.com/azkaban/azkaban/pull/1438?

**Comment by juhoautio** [comment]
@swamisun yes, we have this implemented according to our needs and it's been in production use for a while. For us it was enough for now to implement the bullet points that you can find in the issue description in the beginning of the `Design` section (the ones before `New Job API` etc.)

In our case resumable jobs are typically such that are actually tracking some external task. The external task ids we store into a database in our custom job types so that they can fetch them from there when a job is resumed. It would be nicer if azkaban itself would support `JobState` as described above.

Any way, if this (what we have) suits your needs, I guess I could prepare a pull request to share the code. But currently it seems to me like it won't get to be included in the main Azkaban repository at least any time soon.

**Comment by juhoautio** [comment]
@djaiswal83 DAG rewrite also mentioned here https://github.com/azkaban/azkaban/issues/1127#issuecomment-325883966 – what do you think about this?

**Comment by djaiswal83** [comment]
As I mentioned in other thread, there is no plan to work on it in near future.

**Comment by juhoautio** [comment]
@djaiswal83 thanks. We have automatic resuming implemented in our private fork. Are you open to including it? If yes, I could prepare a pull request. But it’s of course some work, so I wouldn’t be making PR if you don’t see it as a useful contribution.

**Comment by djaiswal83** [comment]
Let me get back on that in a day or two.

**Comment by jakhani** [comment]
@juhoautio This is very useful feature. Thank you so much for working on it. I have couple of doubts. Below I have mentioned the same.

- Which executor will resume the flow? Can resume be picked up by any executor? Or does it go to the same executor?
- It will be good to resume from Dispatch logic place (azkaban.executor.ExecutionController). Please ignore my comment if it is already happening at that place.
- My assumption is : this feature will be useful to treat flow as a stateless entity and resume it from where it is left. It will allow us to resume flows no matter where executors are running, bare metal or containers.
-  I am not clear about resume for streaming or long running jobs. How is it working currently on your production environment?

**Comment by juhoautio** [comment]
> Which executor will resume the flow? Can resume be picked up by any executor? Or does it go to the same executor?

With the current implementation, it may be any of the active executors. 

> It will be good to resume from Dispatch logic place (azkaban.executor.ExecutionController).

Ok, why? In our impl it's done by `RunningExecutionsUpdater` because it monitors running flows and
- it can detect that the executor is not active any more
- it can deduce from a `"Flow does not exist"` response that the execution is not running on the assigned executor any more

When an execution needs to be resumed, `RunningExecutionsUpdater` calls `ExecutorManager`, which then does ` this.queuedFlows.enqueue`.

> My assumption is : this feature will be useful to treat flow as a stateless entity and resume it from where it is left. It will allow us to resume flows no matter where executors are running, bare metal or containers.

In comparison, what would be a stateless flow entity?

> I am not clear about resume for streaming or long running jobs. How is it working currently on your production environment?

For these the Azkaban jobs are tracking some external jobs that remain running continuously, potentially forever. Our streaming job type is such that when the Azkaban job is killed, it terminates the external cluster, and once it has terminated, then the Azkaban job also reaches the KILLED status.

When an execution is moved to another executor, our job type is able to resume tracking the external job because the external job's ID is persisted in a DB and it's read from there when the resumed execution starts. We haven't yet implemented the JobState interface that I specified in the description above, so it's a bit of a hack currently. Having JobState as a feature of Azkaban's job type plugin interface would allow for a clean implementation.

**Comment by arvindpruthi** [comment]
Hi @juhoautio we have similar use case to the resume you mentioned. However, in our case it may not be a forever running streaming job, but jobs that may have been spawned to Yarn.

From 3 years ago, our use cases have evolved and we in fact have started witnessing OOMs. In addition, upgrade/rollback of Executors will indeed be simplified by bringing down the Executor Server and restarting after the upgrade. 

Our (Linkedin's) use case for restart ability will require the flows to get restarted with the same execution id and fast forward to the point in DAG where the FlowRunner was pre-crash. We would also need to connect to all the jobs that were spawned to external entities such as Yarn or Kubernetes. We don't at least for now see the Jobs as "Resumable", but we do need to re-establish a connection with those or restart if required.

Sounds like your implementation hits a number of these points. Therefore, I am really keen to see the PR.

**Comment by juhoautio** [comment]
@arvindpruthi I would like to make a PR. But it would be nice if @djaiswal83 or other repo owner could confirm that there is bandwidth for reviewing & merging it – or are you, @arvindpruthi also eligible for that? :)

> Our (Linkedin's) use case for restart ability will require the flows to get restarted with the same execution id and fast forward to the point in DAG where the FlowRunner was pre-crash. 

Cool, that is how our fork behaves.

> We would also need to connect to all the jobs that were spawned to external entities such as Yarn or Kubernetes. We don't at least for now see the Jobs as "Resumable", but we do need to re-establish a connection with those or restart if required.

To resume & continue tracking external jobs we currently save the external job's id in a custom DB table. Ideally Azkaban's Job interface could natively support saving some kind of job / attempt state that could be read in the job type code on resume. However, I would submit the first version of resuming feature without that building block to keep the scope manageable. Job state could be added after that. Do you have anything against that?

**Comment by arvindpruthi** [comment]
Hi @juhoautio I am one of the repo owners and the tech lead for Azkaban at Linkedin. This is an important PR and busy as we might be, either @djaiswal83 or someone within my team will find time to review the PR. I agree to keep resumability out of the PR for now.

**Comment by juhoautio** [comment]
@arvindpruthi thanks. I'd like to make that PR, but would it be possible to get PR https://github.com/azkaban/azkaban/pull/2515 merged first?

**Comment by djaiswal83** [comment]
@juhoautio we are going through an internal process to manage commits due to which everything is held off. Please be patient as very soon we will bring both internal and external repos to sync and start merging PRs.
